spec/maths/statistics/vector_spec.rb

```diff
@@ -693,24 +693,46 @@
     end
   end
 
-  context "#macd" do
-    it "calculates moving average convergence divergence" do
-      # MACD uses a lot more data than the other ones, so we need a bigger vector
-      data = Daru::Vector.new(
-        File.readlines("spec/fixtures/stock_data.csv").map(&:to_f))
+  describe '#macd' do
+    let(:source) { Daru::DataFrame.from_csv('spec/fixtures/macd_data.csv') }
 
-      macd, signal = data.macd
+    # skip initial records during compare as ema is sensitive to
+    # period used.
+    # http://ta-lib.org/d_api/ta_setunstableperiod.html
+    let(:stability_offset) { 90 }
+    let(:delta) { 0.001 }
 
-      # check the MACD
-      expect(macd[-1]).to be_within(1e-6).of(3.12e-4)
-      expect(macd[-10]).to be_within(1e-4).of(-1.07e-2)
-      expect(macd[-20]).to be_within(1e-5).of(-5.65e-3)
+    let(:macd) { subject[0] }
+    let(:macdsig) { subject[1] }
+    let(:macdhist) { subject[2] }
 
-      # check the signal
-      expect(signal[-1]).to be_within(1e-5).of(-0.00628)
-      expect(signal[-10]).to be_within(1e-5).of(-0.00971)
-      expect(signal[-20]).to be_within(1e-5).of(-0.00338)
+
+    context 'by default' do
+      subject { source['price'].macd }
+
+      it do
+        30.times do |i|
+          idx = stability_offset + i
+          expect(macd[idx]).to be_within(delta).of(source['macd_12_26_9'][idx])
+          expect(macdsig[idx]).to be_within(delta).of(source['macdsig_12_26_9'][idx])
+          expect(macdhist[idx]).to be_within(delta).of(source['macdhist_12_26_9'][idx])
+        end
+      end
```

Comment:
I'd say that testing something in cycle is a really bad tone (because if, for ex, if row 2 fails, it would not be "different in second element", but "expected X, received Y").
What I'd like to see here is something in line with:
```ruby
it { expect { macd }.to be_all_withing(delta).of(source['macd_12_26_9']) }
```
To do this, you'll define your own matcher, it is not hard.

Second thing is those 3 lines (716-718) are not DRY, neither between them, neither comparing with 729-731. I'd say something like this:
```ruby
context 'by default'
 it_should_behave_like 'correct macd', '12_26_9'
end

context 'custom values' do
 it_should_behave_like 'correct macd', '12_26_9', 6, 13, 4
end

# ...and in that shared examples something like:
%w[macd macdsig macdhist].each_with_index do |field, i|
  it { expect(subject[i]).to be_all_withing(delta).of(source["#{field}_#{code}"]) }
end